### PR TITLE
Add support for compile-only dependencies and class path expansion (properties, globs)

### LINKED ISSUE #817: Add support for compile-only dependencies (aka "provided" dependencies)
We should allow users to add "compile-only" dependencies, added to the class path only at compile-time:
```scala
//> using compile-library "com.github.plokhotnyuk.jsoniter-scala::jsoniter-scala-macros:2.13.8"
```

There's already some support for compile-only class path internally (see `Artifacts.compileClassPath`). It should be a matter of adding JARs coming from dependency resolution there.

Implementation-wise, note that we should probably run two dependency resolutions when users add compile-only dependencies: one compile resolution with both compile-only and standard dependencies, and a second one with just standard dependencies.

### LINKED ISSUE #2027: Accept wildcards in --extra-jars
**Is your feature request related to a problem? Please describe.**
`java -cp` accepts `*` characters in the class path, that it expands to all JARs in the directory. (See `java` command line documentation [here](https://docs.oracle.com/javase/6/docs/technotes/tools/windows/classpath.html) for example).

**Describe the solution you'd like**
It'd be nice if `scala-cli` accepted such wildcards too, in `--extra-jars`.

**Describe alternatives you've considered**
Without support for this, users handled a class path value containing such wildcards have to expand these on their own, before passing them to `scala-cli`, which is cumbersome.

### PR DESCRIPTION
**Pull request by alexarchambault**
I've been needing these features in some of the things I'm working on, so I quickly hacked that

### CODE REVIEW

**Comment by alexarchambault** [comment]
I guess the last two commits need a bit of doc…

File: modules/options/src/main/scala/scala/build/Artifacts.scala [hunk 1]
```diff
@@ -356,14 +396,19 @@ object Artifacts {
         .map(_.flatten)
     }
 
+    coursier.Artifacts()
```
Comment (Gedochao) [inline modules/options/src/main/scala/scala/build/Artifacts.scala]: ?

File: modules/options/src/main/scala/scala/build/Artifacts.scala [hunk 2]
```diff
@@ -415,24 +483,31 @@ object Artifacts {
...
+        logger,
+        cache,
+        classifiersOpt,
+        maybeRecoverOnError
+      )
+    }
+    res
+  }
+
+  private def fetchKeepFetcher(
```
Comment (Gedochao) [inline modules/options/src/main/scala/scala/build/Artifacts.scala:499]: This name is a bit confusing, I think... some scaladocs would help, maybe 🤔 
It's hard to get around between `fetch`, `fetch0` and `fetchKeepFetcher`...

File: modules/cli/src/main/scala/scala/cli/exportCmd/MillProjectDescriptor.scala
Comment (Gedochao) [inline modules/cli/src/main/scala/scala/cli/exportCmd/MillProjectDescriptor.scala:1]: we might want some `export` test for compile-only dependencies, for both Mill & SBT

**Comment by Gedochao** [comment]
@alexarchambault What's the current status of this PR? Any plans to finish it in the near future?

**Comment by alexarchambault** [comment]
> What's the current status of this PR? Any plans to finish it in the near future?

I do, likely in the coming weeks, I'd say

**Comment by lwronski** [comment]
> > What's the current status of this PR? Any plans to finish it in the near future?
> 
> I do, likely in the coming weeks, I'd say

@alexarchambault  Hi Alex, I hope you're doing well. I'm just checking to see if you have time to finish this PR soon. If you're busy, I can help to finish it, especially to add support for the compile-only dependencies. Let me know your preference.

**Comment by Gedochao** [comment]
Closing this, as it seems to have been largely delivered already by @lwronski in #2299 (#817).
The remaining requirements are already tracked separately under #2027.

**Pull request closed without merging**